File: solutions/786.cpp
class Solution {
public:
    vector<int> kthSmallestPrimeFraction(vector<int>& arr, int k) {
        int n = arr.size();
        double left = 0;
        double right = 1;
        
        while (true) {
            double mid = left + (right - left) / 2;
            int cnt = 0;
            int p = 0;
            int q = 0;
            int j = 1;
            double maxFrac = 0;
            for (int i = 0; i < n - 1; ++i) {
                while (j < n && arr[i] > mid * arr[j]) j++;
                cnt += (n - j);
                if (j == n) break;
                double f = (double)arr[i] / arr[j];
                if (f > maxFrac) {
                    p = i;
                    q = j;
                    maxFrac = f;
                }
            }
            if (cnt == k) {
                return {arr[p], arr[q]};
            }
            else if (cnt > k) right = mid;
            else left = mid;
        }
        return {};
    }
};

File: rust_solutions/786.rs
impl Solution {
    pub fn kth_smallest_prime_fraction(arr: Vec<i32>, k: i32) -> Vec<i32> {
        let n = arr.len();
        let mut left = 0.0;
        let mut right = 1.0;

        while left < right {
            let mid = left + (right - left) / 2.0;
            let mut count = 0;
            let mut p = 0;
            let mut q = 0;
            let mut j = 1;
            let mut max_frac = 0.0;
            for i in 0..(n - 1) {
                while j < n && (arr[i] as f64) > mid * (arr[j] as f64) {
                    j += 1;
                }
                count += n - j;
                if j == n {
                    break;
                }
                let frac = (arr[i] as f64) / (arr[j] as f64);
                if frac > max_frac {
                    p = i;
                    q = j;
                    max_frac = frac;
                }
            }
            if count as i32 == k {
                return vec![arr[p], arr[q]];
            } else if count as i32 > k {
                right = mid;
            } else {
                left = mid;
            }
        }
        vec![]
    }
}
